lib/tasks/payment_breakdown.rake

```diff
@@ -23,63 +19,35 @@ namespace :payment_calculation do
 
     total_participants = (ARGV[2] || ParticipantEventAggregator.call({ lead_provider: lead_provider })).to_i
     uplift_participants = (ARGV[3] || ParticipantUpliftAggregator.call({ lead_provider: lead_provider })).to_i
-    per_participant_in_bands = lead_provider.call_off_contract.bands.each_with_index.map { |b, i| "£#{b.per_participant.to_i} per participant in #{band_name_from_index(i)}" }.join(", ")
-
-    breakdown = PaymentCalculator::Ecf::PaymentCalculation.call(
-      contract: lead_provider.call_off_contract,
-      total_participants: total_participants,
-      uplift_participants: uplift_participants,
-      event_type: :started,
-    )
-
-    service_fees = breakdown.dig(:service_fees).each_with_object([]) do |hash, bands|
-      bands << [
-        band_name_from_index(bands.length),
-        "£#{number_to_delimited(hash[:service_fee_per_participant].to_i)}",
-        "£#{number_to_delimited(hash[:service_fee_monthly].to_i)}",
-      ]
-    end
-
-    output_payments = breakdown.dig(:output_payments).each_with_object([]) do |hash, bands|
-      bands << [
-        band_name_from_index(bands.length),
-        "£#{number_to_delimited(hash.dig(:started, :per_participant).to_i)}",
-        "£#{number_to_delimited(hash.dig(:started, :subtotal).to_i)}",
-      ]
-    end
+    total_ects = (ARGV[2].present? ? ARGV[2].to_i / 2 : ParticipantEventAggregator.call({ lead_provider: lead_provider, started: :count_active_ects_for_lead_provider }))
+    total_mentors = (ARGV[2].present? ? ARGV[2].to_i - ARGV[2].to_i / 2 : ParticipantEventAggregator.call({ lead_provider: lead_provider, started: :count_active_mentors_for_lead_provider }))
+    Tasks::PaymentBreakdown.new(contract: lead_provider.call_off_contract, total_participants: total_participants, uplift_participants: uplift_participants, total_ects: total_ects, total_mentors: total_mentors).to_table
+  rescue StandardError => e
+    puts e.message
+    puts e.backtrace
+  ensure
+    exit
+  end
 
-    uplift_payment = breakdown[:uplift].each_with_object({}) do |(type, value), hash|
-      hash[type] = "£#{number_to_delimited(value.to_i)}"
+  desc "generate csv payment calculations for all lead providers"
+  task csv: :environment do
+    filename = (ARGV[1] || "output.csv")
+    lead_providers = LeadProvider.all
+    CSV.open(filename, "wb") do |csv|
+      csv << Tasks::PaymentBreakdown.new(contract: lead_providers.first.call_off_contract, total_participants: 0, uplift_participants: 0).csv_headings
+      lead_providers.each do |lead_provider|
+        contract = lead_provider.call_off_contract
```

Comment:
One DRY suggestion for these 6 lines of code as it's repeated in both tasks:

```
def shared(lead_provider)
  total_participants = ParticipantEventAggregator.call({ lead_provider: lead_provider }).to_i
  uplift_participants = ParticipantUpliftAggregator.call({ lead_provider: lead_provider }).to_i
  total_ects = (ARGV[2].present? ? ARGV[2].to_i / 2 : ParticipantEventAggregator.call({ lead_provider: lead_provider, started: :count_active_ects_for_lead_provider }))
  total_mentors = (ARGV[2].present? ? ARGV[2].to_i - ARGV[2].to_i / 2 : ParticipantEventAggregator.call({ lead_provider: lead_provider, started: :count_active_mentors_for_lead_provider }))
  Tasks::PaymentBreakdown.new(contract: lead_provider.call_off_contract, total_participants: total_participants, uplift_participants: uplift_participants, total_ects: total_ects, total_mentors: total_mentors)
end
```

and then:

```
shared(lead_provider).to_table
```

```
lead_providers.each do |lead_provider|
  csv << shared(lead_provider).csv_body
end
```

Reply:
Actually, I'm going to make those blocks slightly different. The 'csv' option shouldn't take in ARGV[2], since it should always be hitting the database.

Reply:
Also. Having a method in the rake task gives it global scope, which isn't a good thing, since all rake tasks are loaded and can redefine. For reference:
https://kevinjalbert.com/defined_methods-in-rake-tasks-you-re-gonna-have-a-bad-time/